### ISSUE TITLE
Custom error pages (and captcha support)

### ISSUE THREAD

**Issue by Simbiat** [hslatman/caddy-crowdsec-bouncer#46]
My [config](https://github.com/Simbiat/simbiat.ru/blob/master/config/frankenphp/caddy.json5) has custom error pages, but looks like they are not being applied if access is blocked by CrowdSec (they do get applied if it's done by Coraza, though). Instead I am seeing generic Caddy's page (not the one from CrowdSec either). Furthermore if I block IP with "captcha" instead of "ban" I am still getting 403 error, instead of expected 401.
Is this some limitation of the plugin or am I doing something wrong?

**Comment by LaurenceJJones** [comment]
Seems captcha is a TODO:

https://github.com/hslatman/caddy-crowdsec-bouncer/blob/6022901f3397e5fe5d01f744a23986754746eabe/http/http.go#L127-L130

There is no custom pages, so this needs to be implemented also as since crowdsec directive acts as a middleware it stops the chaining input. I dont know how the coraza implementation works to get around this.

**Comment by hslatman** [comment]
That's quite the config, @Simbiat 😛 

Yes, captcha has been on my list for a while now. Main thing to decide is what technology to use, as I don't want to force people to use reCaptcha, for example. Ideally I would like to make it pluggable.

Without looking much deeper into the code, I think this might be what makes the error pages work with Coraza: https://github.com/corazawaf/coraza-caddy/blob/main/coraza.go#L116-L131.

**Comment by LaurenceJJones** [comment]
> Without looking much deeper into the code, I think this might be what makes the error pages work with Coraza: https://github.com/corazawaf/coraza-caddy/blob/main/coraza.go#L116-L131.

Ahhh so returning that type / struct informs caddy to trigger the pages it has in it directives interesting.

**Comment by Simbiat** [comment]
If you strip forced MIME types from my config it will be much more tame 😅
I can confirm, that Coraza seems to return something about "Interruption", which I can get from error messages in Caddy, which matches `errInterruptionTriggered`. So, I guess it's just required to return `caddyhttp.HandlerError` then. If that's the case, then it could solve the "pluggable" CAPTCHA, too: return 401 for CAPTCHA and 403 for everything else, and let users handle 401 with whatever CAPTCHA thing they want. The only other thing that would be useful here is to have some variable to indicate that 403/401 was caused by CrowdSec, and ideally some details on what triggered the interruption (same as what I've asked for for Coraza module [here](https://github.com/corazawaf/coraza-caddy/issues/162)).

**Comment by hslatman** [comment]
> I can confirm, that Coraza seems to return something about "Interruption", which I can get from error messages in Caddy, which matches `errInterruptionTriggered`. So, I guess it's just required to return `caddyhttp.HandlerError` then. If that's the case, then it could solve the "pluggable" CAPTCHA, too: return 401 for CAPTCHA and 403 for everything else (optionally allow customization), and let users handle 401 with whatever CAPTCHA thing they want.

I can see that working, but I think you'll still want that to be connected to the fact that it was CrowdSec that decided a specific action was needed, vs. some other reason a 401 was returned.

> The only other thing that would be useful here is to have some variable to indicate that 403/401 was caused by CrowdSec, and ideally some details on what triggered the interruption (same as what I've asked for for Coraza module [here](https://github.com/corazawaf/coraza-caddy/issues/162)).

Yes, I think this is useful, and also ties in with my previous line. I'll see what I can do 🙂

**Comment by LaurenceJJones** [comment]
> I can see that working, but I think you'll still want that to be connected to the fact that it was CrowdSec that decided a specific action was needed, vs. some other reason a 401 was returned.

I agree with @hslatman Captcha is alot more involved, you need to track the request / responses from the Captcha itself rather than looking just for `401` since you need to know when to validate them with the provider. It's something similar I have done on [HAProxy SPOA](https://github.com/LaurenceJJones/crowdsec-spoa) which is using a session cookie which has an internal state.

**Comment by Aetylus** [comment]
I don't know how idiomatic or kosher this would be in regards to implementing CrowdSec's `captcha` decision, but would implementing [Anubis](https://github.com/TecharoHQ/anubis) or something similar be feasible?

At a high level, without looking into it in more depth myself, it feels like it could be simpler because you don't need to track a captcha response back. And I would think that, in general, if the idea of a captcha is to distinguish between a possible bot and a human, then something designed to prevent bots from scraping content would at some level fulfill a similar enough purpose?

This is more food for thought that might be worth looking into rather than a suggestion.

**Comment by LaurenceJJones** [comment]
> I don't know how idiomatic or kosher this would be in regards to implementing CrowdSec's `captcha` decision, but would implementing [Anubis](https://github.com/TecharoHQ/anubis?rgh-link-date=2025-05-27T03%3A08%3A50.000Z) or something similar be feasible?
> 
> At a high level, without looking into it in more depth myself, it feels like it could be simpler because you don't need to track a captcha response back. And I would think that, in general, if the idea of a captcha is to distinguish between a possible bot and a human, then something designed to prevent bots from scraping content would at some level fulfill a similar enough purpose?
> 
> This is more food for thought that might be worth looking into rather than a suggestion. I did come across an blog post [here](https://fabulous.systems/posts/2025/05/anubis-saved-our-websites-from-a-ddos-attack/) that reports Anubis as successfully preventing a DDoS as well, which feels like at least part of the reason one might want to have a captcha to begin with.

Yes if your aim is to filter between bots and humans then Anubis is probably the better candidate as it screens _all_ users. However, if your aim is to only screen malicious IPs then Anubis cant check our decisions and only impose on those requests / users.

I havent looked much into Anubis but AFAIK you would have to spin up one per application as it meant to be a proxy in between your reverse proxy and the application so little more overhead but I have seen it quite performant and easy to use.

**Comment by Simbiat** [comment]
@LaurenceJJones, I think @Aetylus meant to suggest, that if Crowdsec captcha is supposed to be triggered - pass it to Anubis. Which , probably, could make sense, it's like another layer of protection, after all. But that would require either a separate Caddy module (which was requested in https://github.com/TecharoHQ/anubis/issues/16 and is being worked in https://github.com/daegalus/caddy-anubis) or an integration with Anubis on Crowdsec side. I think neither is relevant to the bouncer module, though, at least not directly.

**Comment by LaurenceJJones** [comment]
> [@LaurenceJJones](https://github.com/LaurenceJJones), I think [@Aetylus](https://github.com/Aetylus) meant to suggest, that if Crowdsec captcha is supposed to be triggered - pass it to Anubis. Which , probably, could make sense, it's like another layer of protection, after all. But that would require either a separate Caddy module (which was requested in [TecharoHQ/anubis#16](https://github.com/TecharoHQ/anubis/issues/16) and is being worked in [daegalus/caddy-anubis](https://github.com/daegalus/caddy-anubis?rgh-link-date=2025-05-27T07%3A56%3A44.000Z)) or an integration with Anubis on Crowdsec side. I think neither is relevant to the bouncer module, though, at least not directly.

Those issues are asking to use Anubis as a module so you dont have to run an external anubis as a proxy as I understand. The goal of Anubis is you send all traffic to it so it can screen every user, as then you would have the most chance to catch bots as only picking and choosing which traffic to send means you could get false negatives.

**Comment by Aetylus** [comment]
@Simbiat is correct in what I was thinking about. While I think the idea of Anubis is to have it on your proxy for _all_ requests, the request flow I'm more envisioning is:

1. Request comes into Caddy
2. Caddy CrowdSec module parses the request and sends it to the LAPI to get a decision
3. Caddy bouncer handles decision
3a. If `ban` decision, Caddy handles the ban request as the module already functions
3b. If `captcha` decision, Caddy forwards request to *some* proof-of-work, whether that's Anubis or something like the [Cerberus](https://github.com/sjtug/cerberus) module mentioned in https://github.com/TecharoHQ/anubis/issues/16
4b. Anubis/Cerberus/some PoW module returns request to the rest of the route flow(?) or handles forwarding the request onwards to its destination itself

This is overly simplified, and this does in fact circumvent in some ways the purpose of Anubis by only directing some traffic to it rather than all. In this case, the PoW is acting as a faux captcha in the sense that it's functioning as a system to deter bots but isn't actually a captcha itself.

**Comment by Simbiat** [comment]
Personally, I think it should be fine to have Anubis here as a layer to analyze only part of the requests. If I understand correctly, Crowdsec bans based on known bad IPs and known patterns, "known" being the keyword. But PoW (and Captcha) is meant more for *suspicious behavior* which may be malicious or may be not. If Crowdsec blocks the request we already know that it's a bad one, if not, then we may want to check if through PoW logic to determine if it is legitimate, and if not - block it. That way we should waste less resources on PoW logic, since we are not covering everything.

But anyway, I do not think this is relevant to this issue or to Crowdsec/Appse implementation in a Caddy module. I think it's more of a question of what we do with failed requests, and that should be handled in #77. This issue is more about native captcha support, since Crowdsec does have that functionality in it, if I understand correctly. Perhaps I should rename the issue to make it a bit clearer.

On the other hand, it's possible that implementing proper bubbling and sharing of Crowdsec decision details would make this issue obsolete entirely, because you will be able to then process the request using whatever other module, and you can watch for ban/captcha in the decision, I guess.

**Comment by Aetylus** [comment]
Oh my mistake! I think part of this line of thought was I didn't realize CrowdSec had native captcha support to begin with, and I was under the assumption that there were some questions about how to implement a captcha into the bouncer.

**Comment by hslatman** [comment]
It could be an option to use Anubis as the captcha implementation when a `captcha` decision is active for an IP. But I'd still like it to be pluggable, in the sense that I don't want to rely on it solely. 

Implementation wise it could depend on the Anubis Caddy module, and 1) if it's compiled into the binary, 2) the captcha implementation is configured to use Anubis, and 3) a `captcha` decision is active for an IP, it would go through Anubis (instead of a different captcha implementation).

**Comment by DenuxPlays** [comment]
Hey @hslatman ,

I plan to deploy your caddy bouncer at work but captcha support the only thing that is currently missing.
I would try to implement it but do you have an overview of how you visioned this to be implemented and/or what needs to be done?

**Comment by Simbiat** [comment]
While working on #125, I looked into what is needed for Captcha processing, now that we can have custom error pages. In fact, I decided to update how the errors are returned specifically to be able to handle Captcha differently from Ban, since #107 was setting identical error message for both, and thus you would not be able to differentiate between the two neither in `errors` route, nor in PHP or other handler. But I am a bit confused by how this is supposed to work even, and I feel like I am missing something.

My understanding of how it _should_ be handled in general is:
1. A bouncer receives a `captcha` response.
2. It routes the response to a custom page, which presents some sort of a "challenge" (reCAPTCHA or a round of chess, should not matter).
3. Once user solves the "challenge", something needs to change to the "state" of the user session, so that next request is not blocked.  Looks like usually that's something added to sessional cookie.
4. When the next request is issued, the bouncer sees that "proof of the challenge completion" and...
5. Disregards the block from `CrowdSec`.
6. If an `AppSec` block is issues here, it's supposed to be processed normally.

Is my understanding correct, that  we do not remove the decisions? https://crowdsecurity.github.io/api_doc/lapi/#/ does show that there is a `DELETE` method supported, but required a JWT, and is meant for "watchers", not bouncers. "Watchers" are `cscli` and "CrowdSec Agents", and I was not able to find a clear definition for the latter. But point is that bouncers can't remove decisions (that's fine), and even if they could, they can't remove one using an ID, meaning, that if they could remove them, they could accidentally remove more than 1 decision, which may not be what we want.

But ultimately, is it that if I want to have a custom Captcha processing, I can do a custom error page with a button, clicking which creates an AJAX request to which a header like `X-Captcha-Passed` is added and then in Caddy config instead of
```json
"routes": [
  {
    "handle": [
      {
        "handler": "crowdsec"
      }
    ]
  },
  {
    "handle": [
      {
        "handler": "appsec"
      }
    ]
  }
]
```
I have
```json
"routes": [
  {
    "handle": [
      {
        "handler": "crowdsec"
      }
    ],
    "match": [
      {
        "header": {
          "X-Captcha-Passed": null
        }
      }
    ]
  },
  {
    "handle": [
      {
        "handler": "appsec"
      }
    ]
  }
]
```
which will trigger CrowdSec only if the header is missing. Of course the above is just a simple example, that's obviously not a secure way. But am I understanding correctly, that this is the way things are supposed to work like? And why can't we remove the decision directly from bouncer, if we have its ID? @LaurenceJJones, perhaps you can clarify this a bit?

**Comment by LaurenceJJones** [comment]
Bouncers cannot write using the api key they only have read access, plus you have the further issue of NAT IE a user behind a NAT can solve the captcha but how do you allow only valid users and not blanket the IP itself.

So the way we have done it in most first party bouncers is they either have an internal cache of "valid ips" (nginx bouncer) that once the captcha is solved the IP is added to an internal cache and it check "if ip in cache" then allow (pending appsec checks). However, this is full of holes EG the NAT stuff above.

So what I did for [HAProxy SPOA](https://github.com/crowdsecurity/cs-haproxy-spoa-bouncer) instead was use a signed JWT cookie which has a configure value for a secret that if you are LBing or whatever then you can pass the same secret across instances then whatever instance checks the signed JWT it will be valid for. This enabled us to do a per device which in theory is "per user" captcha checking, which was way way better than blanketing the IP.

so _tldr_ is you dont remove the decision, you just have a internal cache of "solved" things or use a signed JWT cookie value which gets checked whenever the user gets a captcha enforcement. (which nowadays the jwt approach is the way to go as that scales better in lb / HA environments)

so it kind of looks like this flow

```mermaid
flowchart TD
    A[Request hits bouncer] --> B{Captcha enforcement decision?}

    B -->|No| C[Allow request]

    B -->|Yes| D{Valid JWT cookie exists?}

    D -->|No| E[Issue signed JWT cookie]
    E --> E1["Claims: solved=false, exp, nonce/session id"]
    E1 --> F[Show captcha]

    D -->|Yes| G{JWT signature valid?}
    G -->|No| E

    G -->|Yes| H{solved == true?}
    H -->|Yes| I[Allow request pending AppSec checks]
    H -->|No| F

    F --> J[User solves captcha]
    J --> K{Captcha verification passed?}

    K -->|No| F
    K -->|Yes| L[Create new JWT]
    L --> L1["Claims: solved=true, exp, nonce/session id"]
    L1 --> M[Sign JWT using configured secret]
    M --> N[Set updated JWT cookie]

    N --> O[Next request includes solved JWT]
    O --> D
```

also im not apart of CrowdSec anymore, happy to help but I cant aid as in influence the team for things 👍🏻

**Comment by Simbiat** [comment]
I see. But technically that does not even have to be done by a bouncer, you can use something else, and then use that in a `match` clause in Caddy config, correct?

**Comment by LaurenceJJones** [comment]
> I see. But technically that does not even have to be done by a bouncer, you can use something else, and then use that in a `match` clause in Caddy config, correct?

I have 2% knowledge of Caddy and the configuration all I know is the crowdsec side and how we done it in other bouncers 🤷🏻

**Comment by hslatman** [comment]
> I see. But technically that does not even have to be done by a bouncer, you can use something else, and then use that in a `match` clause in Caddy config, correct?

If you still want to involve the decision from CrowdSec, I don't see how you're not using a bouncer in some way. 

It involves some kind of additional state that the bouncer will need to put in place, and to check it again. Or delegate that responsibility to another thing, but I believe that'll only make things more complex.

**Comment by Simbiat** [comment]
I mean, if we need to check some cookie, then it can be
```json
"routes": [
  {
    "handle": [
      {
        "handler": "crowdsec"
      }
    ],
    "match": [
      {
        "header": {
          "X-Captcha-Passed": null
        }
      }
    ]
  },
  {
    "handle": [
      {
        "handler": "appsec"
      }
    ]
  }
]
```
but with something else besides `"X-Captcha-Passed": null` or some different `match` logic.

**Comment by hslatman** [comment]
What I'm trying to say is that I think it'll be more complex if it comes down to having to put it in different places in the configuration. While I can see it work, I think it'll be simpler to understand if the bouncer handles the whole thing.
